### PR TITLE
Added weighted loss

### PR DESCRIPTION
**Pull request by DuncanZauss**

### CODE REVIEW

**Comment by svenkreiss** [comment]
@cla-bot recheck

**Comment by svenkreiss** [comment]
Thanks! @cla-bot[bot] check

File: openpifpaf/network/losses/composite.py [hunk 1]
```diff
@@ -82,15 +87,20 @@ def _confidence_loss(self, x_confidence, t_confidence):
         ce_loss = self.confidence_loss(x_confidence, bce_target)
         if self.prescale != 1.0:
             ce_loss = ce_loss * self.prescale
+        weights_torch = torch.tensor(self.weights, device=bce_masks.device).unsqueeze(0).unsqueeze(2).unsqueeze(3)  # TODO find a nicer way to assign the weights to the correct GPU
```
Comment (svenkreiss) [inline openpifpaf/network/losses/composite.py:90]: This tensor is the same in every forward pass. Please move it to the constructor and use it in place of `self.weights`. Once that is done, you don't need to assign the `device` anymore. I would first create a tensor initialized to ones (`torch.ones`) with shape `[1, n_fields, 1, 1]`. Then if `w` is not None, you set the entries with something like `self.weights[0, :, 0, 0] = w`.

**Review by svenkreiss — CHANGES_REQUESTED** [review]
Thanks Duncan. I have added one comment below and then the linters have some suggestions :)

File: openpifpaf/network/losses/composite.py [hunk 2]
```diff
@@ -31,7 +31,12 @@ def __init__(self, head_net: heads.CompositeField3):
             + ['{}.{}.scales{}'.format(head_net.meta.dataset, head_net.meta.name, i + 1)
                for i in range(self.n_scales)]
         )
-
+        w = head_net.meta.weights
+        n_fields=head_net.meta.n_fields
+        self.weights = torch.ones([1, n_fields, 1, 1])
```
Comment (svenkreiss) [inline openpifpaf/network/losses/composite.py:36]: Ok, sorry, I am going a bit back on what I said before. To preserve behavior to previous unweighted case:
Initialize `self.weights = None` and only set it when a w is given. In the forward functions, you then only apply the weight when `self.weights is not None`. Then the behavior is exactly the same as before.

File: openpifpaf/network/losses/composite.py [hunk 3]
```diff
@@ -82,6 +87,10 @@ def _confidence_loss(self, x_confidence, t_confidence):
         ce_loss = self.confidence_loss(x_confidence, bce_target)
         if self.prescale != 1.0:
             ce_loss = ce_loss * self.prescale
+        weight = torch.ones_like(t_confidence, dtype=torch.float, requires_grad=False)
```
Comment (svenkreiss) [inline openpifpaf/network/losses/composite.py:90]: Here you don't need to specify dtype. It will obtain it from t_confidence. Remove it and then it will do the right thing no matter whether you are working with float32 or float64.

File: openpifpaf/network/losses/composite.py [hunk 4]
```diff
@@ -104,13 +112,16 @@ def _localization_loss(self, x_regs, t_regs):
             )
             if self.prescale != 1.0:
                 loss = loss * self.prescale
+            weight = torch.ones_like(reg_masks, dtype=torch.float, requires_grad=False)
```
Comment (svenkreiss) [inline openpifpaf/network/losses/composite.py:115]: Here also, remove the explicit dtype. However, the mask will have the wrong dtype. You can use `t_regs[:, :, 0]` instead.

File: openpifpaf/network/losses/composite.py [hunk 5]
```diff
@@ -121,6 +132,10 @@ def _scale_losses(self, x_scales, t_scales):
             )
             if self.prescale != 1.0:
                 loss = loss * self.prescale
+            weight = torch.ones_like(mask, dtype=torch.float, requires_grad=False)
```
Comment (svenkreiss) [inline openpifpaf/network/losses/composite.py:135]: similarly, use `t_scales[:, :, 0]` and remove explicit dtype.

File: openpifpaf/network/losses/composite.py (hunk 2 above)
Comment (svenkreiss) [inline openpifpaf/network/losses/composite.py:36]: just to be absolutely sure, also use `requires_grad=False` here.

File: openpifpaf/network/losses/composite.py [hunk 6]
```diff
@@ -92,6 +103,10 @@ def _localization_loss(self, x_regs, t_regs):
         batch_size = t_regs.shape[0]
 
         reg_losses = []
+        if self.weights is not None:
+            weight = torch.ones_like(t_regs[:, :, 0], requires_grad=False)
+            weight[:] = self.weights
+            weights_stacked = weight.unsqueeze(0).repeat(self.n_vectors, 1, 1, 1, 1)
```
Comment (svenkreiss) [inline openpifpaf/network/losses/composite.py:109]: Hi @DuncanZauss , I saw you are cleaning things up. Something must have worked :)
While you are cleaning up, can you remove this `repeat`? From the idea of "Composite" fields, there should only ever be one consistent weight across all the components.

In the language of the paper:
image
We want to weigh the training of this object p with one weight per J.

Comment (DuncanZauss) [reply openpifpaf/network/losses/composite.py:109]: Hi Sven,
I created this repeat in case the weighing is also used for fields, where there are multiple vector/scales, e.g. if a CAF should be weight. I was unsure if the autograd would still work if I use the same weight for all vector/scale fields.

File: openpifpaf/network/losses/composite.py [hunk 7]
```diff
@@ -104,6 +119,8 @@ def _localization_loss(self, x_regs, t_regs):
             )
             if self.prescale != 1.0:
                 loss = loss * self.prescale
+            if self.weights is not None:
+                loss = loss * torch.masked_select(weights_stacked[i, :, :, :, :], reg_masks)
```
Comment (DuncanZauss) [inline openpifpaf/network/losses/composite.py:123]: So I meant inside this for loop, I was unsure if it is okay for the autograd to use the same weight for different vector fields.

Comment (svenkreiss) [reply openpifpaf/network/losses/composite.py:123]: There are actually two reasons why this is ok to remove:
1) the weights don't have gradients themselves. For Autograd, this is just constant.
2) Module tree != computation graph. One of the magic things that goes on behind the scenes. Think about LSTM cells or anything like this where the same linear layer is re-used over and over with different inputs before `backwards()` is called. PyTorch "unrolls" these operations in the computation graph. This is one of the reasons why training can take so much more GPU memory than prediction.

Comment (DuncanZauss) [reply openpifpaf/network/losses/composite.py:123]: Nice, thank you for that thorough explanation! I will change it now.

File: openpifpaf/headmeta.py [hunk 8]
```diff
@@ -48,6 +48,8 @@ class Cif(Base):
     decoder_min_scale = 0.0
     decoder_seed_mask: List[int] = None
 
+    weights: List[float] = None
```
Comment (svenkreiss) [inline openpifpaf/headmeta.py:51]: Sorry, one more request (naming is important), please change this to `training_weights`.

**Comment by svenkreiss** [comment]
Excellent! Done 🥳

**Review by svenkreiss — APPROVED** [review]
